lib/propolis/src/vmm/time_adjust.rs

```diff
@@ -0,0 +1,348 @@
+    )]
+    GuestUptimeOverflow {
+        /// source host hrtime
+        src_hrt: i128,
+
+        /// input guest boot_hrtime
+        boot_hrtime: i128,
+    },
+
+    /// Invalid total delta for boot_hrtime calculations
+    #[error(
+        "could not calculate time delta:\
+            guest uptime {uptime:?}, migrate_delta={migrate_delta:?}"
+    )]
+    TimeDeltaOverflow {
+        /// guest uptime
+        uptime: Duration,
+
+        /// migration time delta
+        migrate_delta: Duration,
+    },
+
+    /// Invalid calculated boot_hrtime
+    #[error(
+        "guest boot_hrtime cannot be represented:\
+            desc=\"{desc}\", total_delta={total_delta:?},\
+            boot_hrtime={boot_hrtime}, dst_hrtime={dst_hrtime:?}"
+    )]
+    BootHrtimeOverflow {
+        /// error description
+        desc: String,
+
+        /// calculated total delta (uptime + migration delta)
+        total_delta: Duration,
+
+        /// input guest boot_hrtime
+        boot_hrtime: Hrtime,
+
+        /// destination host hrtime
+        dst_hrtime: Duration,
+    },
+
+    /// Invalid guest TSC adjustment
+    #[error(
+        "could not calculate TSC adjustment:\
+            desc=\"{desc:?}\", migrate_delta={migrate_delta:?},
+            guest_hz={guest_hz}, tsc_adjust={tsc_adjust}"
+    )]
+    TscAdjustOverflow {
+        /// error description
+        desc: String,
+
+        /// migration time delta
+        migrate_delta: Duration,
+
+        /// guest TSC frequency (hz)
+        guest_hz: u64,
+
+        /// calculated TSC adjustment
+        tsc_adjust: u128,
+    },
+
+    /// Guest TSC overflow
+    #[error("guest TSC overflow: old tsc = {tsc}, adjustment = {adjust}")]
+    GuestTscOverflow {
+        /// input guest TSC
+        tsc: u64,
+
+        /// calculated TSC adjustment
+        adjust: u64,
+    },
+}
+
+/// Returns a non-atomic snapshot of the current hrtime and wall clock time.
+// TODO: We may want to put a bound in here to make sure the two
+// values are close enough, and we may want that to be tunable. What's the
+// best tuning mechanism here? And what should the bound be?
+// We could also just punt on this for now.
+pub fn host_time_snapshot() -> Result<HostTime, TimeAdjustError> {
+    let hrtime;
+    let wall_clock;
+
+    match time::get_highres_time() {
+        Ok(hrt) => {
+            hrtime = hrt;
+        }
+        Err(e) => {
+            return Err(TimeAdjustError::HighResClock(e.to_string()));
+        }
+    };
+
+    match time::get_real_time() {
+        Ok(wc) => {
+            wall_clock = wc;
+        }
+        Err(e) => {
+            return Err(TimeAdjustError::RealTimeClock(e.to_string()));
+        }
+    }
+
+    Ok(HostTime { hrtime, wall_clock })
+}
+
+/// Find the perceived wall clock time difference between when timing data
+/// was read on the source versus the current time on the destination.
+// TODO(#357): Right now we throw up our hands if the delta is negative.
```

Comment:
Could we avoid this problem by trying to measure elapsed migration time from the point of view of just one of the actors involved?

That is, suppose the Pause phase included some kind of mutual acknowledgement in the protocol:

- Source sends an "I'm pausing" message
- Target records the time it received this message and immediately sends an "OK to pause" message
- Only then does the source start pausing

(This could also be an "I read timing data"/"OK you read the timing data" message pair if that's what we want; I'll probably work it out as I keep reviewing this.)

This can produce a "total migration time" estimate on the target that's always *at least* the amount of time the VM was actually paused. It might be an overestimate because it also includes the time it took for the source to receive the "OK to pause now" message and actually start pausing things. We could possibly correct for that further by having the *source* measure some length of time (e.g. from when it sent its "am I OK to pause?" message to when it actually paused things) and then send that to the target to use to correct its view of the total migration time.

Anyway: I haven't totally worked out the arithmetic here, so I could be totally wrong about all this, but I'd be interested in seeing if there's an approach that doesn't rely on real time synchronization between the two machines involved.

Reply:
Using the wallclock difference (and assuming the respective clocks on either side are adequately synced via NTP) seemed like a simple way to get this off the ground.  Ultimately we want some means of judging the passage of time (from some global observer) between the point at which the timing measurement (TSC, boot_hrtime, etc) is taken on the source, and when its adjusted values are to be loaded on the destination.

It's been suggested that we could do more sophisticated things here in the future, including something like a little mini implementation of the NTP algo between the two propolis instances, but cheating with wallclock in the short term seemed like a reasonable starting point.

Reply:
FWIW, the problem I ran into with a negative wall clock delta was in an earlier prototype where the time data was sent immediately after collecting it. In practice (if not for longer delay between reading and sending, but hopefully also because ntp is better behaved than on the lab machines I was on), my hope is that this won't matter, and I think I'm comfortable with leaving this behavior for now. When we begin to optimize things such that migrations are faster, we may want to reconsider it (and I'm happy to leave #357 as a placeholder for that).

Reply:
In d09938fb, which has a much tighter window between export and import than previous versions of this PR, I clamped the migrate delta to 0 if for some reason we see a negative delta.